presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -103,6 +121,26 @@ public LookupSourceFactory compileLookupSourceFactory(List<? extends Type> types
         }
     }
 
+    public JoinFilterFunctionVerifierFactory compileJoinFilterFunctionVerifierFactory(Optional<JoinFilterFunction> joinFilterFunction)
```

Comment:
I think the problem with that approach is that currently `filterFunction` instance is constructed in `LocalExecutionPlanner.createLookupJoinSource`. And we do not have the build pages available at this point. So we cannot just replace `filterFunction` with `filterFunctionVerifier` instance here.
We need to either introduce `JoinFilterFunctionVerifierFactory` or pass build pages as parameter on each call to applyFilter.
I am not super convinced that would be strictly better.

Anyway. I will apply other comments first and leave this as separate final refactoring.